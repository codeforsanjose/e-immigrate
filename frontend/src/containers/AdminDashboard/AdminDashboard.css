.AdminDashboard {
    margin: 0 10%;
}
.AdminDashboard .loading {
    position: absolute;
    z-index: 10;
    font-size: 40px;
    color: white;
    background-color: rgba(0, 0, 0, 0.25);
    width: 90%;
    height: 90%;
    text-align: center;
    margin: 0 auto;
}

.AdminDashboard .responses {
    /* border: solid 1px #efefef; */
    display: grid;
}

.AdminDashboard table {
    border-collapse: collapse;
    table-layout: auto;
    max-height: 600px;
    overflow-y: scroll;
}
.AdminDashboard .responses tbody tr .contact-with-police {
    text-align: left;
    font-size: 0.7em;
    width: 200px;
}
.AdminDashboard .responses tbody tr .contact-with-police-explain {
    text-align: left;
    font-size: 0.7em;
    width: 300px;
}
.AdminDashboard .responses tbody tr td,
.AdminDashboard .responses tbody tr th {
    border-bottom: solid 1px #efefef;
    height: 32px;
    overflow-x: scroll;
}

.AdminDashboard .responses th:nth-child(1),
.AdminDashboard .responses td:nth-child(1) {
    min-width: 40px;
}

.AdminDashboard .responses th:nth-child(2),
.AdminDashboard .responses td:nth-child(2) {
    min-width: 52px;
}

.AdminDashboard .responses th:nth-child(3),
.AdminDashboard .responses td:nth-child(3) {
    min-width: 96px;
}

.AdminDashboard .responses th:nth-child(4),
.AdminDashboard .responses td:nth-child(4) {
    min-width: 104px;
}

.AdminDashboard .responses th:nth-child(5),
.AdminDashboard .responses td:nth-child(5) {
    min-width: 120px;
}

.AdminDashboard .responses th,
.AdminDashboard .responses td {
    padding: 8px;
    text-align: center;
}

.AdminDashboard .responses th {
    color: var(--color-primary);
}

.AdminDashboard .responses th svg {
    margin-left: 6px;
    cursor: pointer;
}

.AdminDashboard .responses th .up {
    transform: rotate(180deg);
}

.AdminDashboard .checkbox {
    width: 20px;
    height: 20px;
}

.AdminDashboard .all-answers {
    display: grid;
    grid-gap: 16px;

    grid-auto-flow: column;
    grid-auto-columns: minmax(1fr, 160px);
    width: 700px;
    overflow-x: scroll;
}
.AdminDashboard .all-answers .answer {
    width: 100%;
    padding: 6px;
    align-self: center;
    min-height: 40px;
}

.AdminDashboard .green {
    background-color: var(--color-green);
}
.AdminDashboard .red {
    background-color: var(--color-red);
}

.AdminDashboard .flag {
    width: 20px;
    height: 20px;
    border-radius: 25px;
    margin: auto;
}
.AdminDashboard .green-outline {
    border: solid 3px var(--color-green);
}
.AdminDashboard .red-outline {
    border: solid 3px var(--color-red);
}

.AdminDashboardModal .Modal .modal-details {
    overflow-y: scroll;
}
.AdminDashboardModal .Modal .green-outline {
    border: solid 3px var(--color-green);
}
.AdminDashboardModal .Modal .red-outline {
    border: solid 3px var(--color-red);
}
.AdminDashboardModal {
    article {
        margin: 6px;
        padding: 6px;
        border-radius: 25px;
    }
}

.AdminDashboard .dropdown-select {
    margin: 0.5rem;
}
.AdminDashboard .dashboard-section-title {
    font-weight: 800;
    margin: 2rem 0 1rem;
}

.AdminDashboard .dashboard-card {
    box-shadow: 0px 2px 4px rgba(0, 0, 0, 0.25), 0px -2px 4px rgba(0, 0, 0, 0.1);
    border-radius: 4px;
    padding: 16px;
    height: 148px;
}

.AdminDashboard .flag-dashboard-card {
    display: grid;
    grid-template-columns: auto;
    grid-template-rows: auto auto auto auto;
    gap: 4px;
}

.AdminDashboard .flag-dashboard-card div {
    display: flex;
    justify-content: space-between;
}

.AdminDashboard .dashboard-card .text-red {
    color: var(--color-red);
}

.AdminDashboard .dashboard-card .text-green {
    color: var(--color-green);
}

.AdminDashboard .dashboard-card .bold {
    font-weight: 700;
}
.AdminDashboard .dashboard-card h4 {
    font-weight: 600;
    font-size: 17px;
    margin: 0;
}

.AdminDashboard .dashboard-buttons-container {
    margin-top: 32px;
    display: flex;
    grid-template-columns: auto auto;
    grid-template-rows: auto;
    gap: 16px;
}

.AdminDashboard .overview-container {
    display: grid;
    grid-template-columns: 200px 412px;
    grid-template-rows: auto;
    gap: 50px;
}

.AdminDashboard .overview-container .agency-grid {
    display: grid;
    grid-template-columns: repeat(auto-fit, 150px);
    grid-template-rows: repeat(3, auto);
    column-gap: 76px;
    row-gap: 4px;
    margin-top: 4px;
    column-rule: 1px solid black;
}

.AdminDashboard .overview-container .agency-grid .agency-row {
    display: grid;
    grid-template-columns: 45px repeat(3, 35px);
    grid-template-rows: auto;
}

.AdminDashboard .overview-container .agency-grid .agency-row .sum {
    text-align: end;
}

.AdminDashboard .delete {
    width: 20px;
    height: 20px;
    border-radius: 25px;
    margin: auto;
    background: #fa6900;
    color: #fff;
    opacity: 0.3;
}

.AdminDashboard .delete:hover {
    opacity: 1;
}
