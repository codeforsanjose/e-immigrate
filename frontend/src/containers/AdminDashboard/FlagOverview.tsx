import React from 'react';
import { QuestionnaireResponseElement } from './types';
type FlagOverviewProps = {
    questionnaireResponses: Array<QuestionnaireResponseElement>;
};
export function FlagOverview(props: FlagOverviewProps) {
    const {
        questionnaireResponses,
    } = props;
    return (
        <div className="flag-dashboard-card">
            <h4>Responses</h4>
            <div>
                <span>Red:</span>
                <span className="text-red bold">
                    {questionnaireResponses.filter(
                        (response) => response.flag === true,
                    ).length}
                </span>
            </div>
            <div>
                Green:{' '}
                <span className="text-green bold">
                    {questionnaireResponses.filter(
                        (response) => response.flag === false,
                    ).length}
                </span>
            </div>
            <div>
                Total:{' '}
                <span className="bold">
                    {questionnaireResponses.length}
                </span>
            </div>
        </div>
    );
}
