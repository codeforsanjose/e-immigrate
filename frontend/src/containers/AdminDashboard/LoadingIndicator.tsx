import React from 'react';
type LoadingIndicatorProps = {
    loading: boolean;
};
export function LoadingIndicator(props: LoadingIndicatorProps) {
    const {
        loading,
    } = props;
    if (!loading) return null;
    return (
        <div className="loading is-vcentered">Loading...</div>
    );
}