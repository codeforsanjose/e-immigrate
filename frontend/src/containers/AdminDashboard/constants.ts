import { ArrayElementOf } from '../../types/ArrayElementOf.js';


export const DESCRIPTIVE_TIMESTAMP = 'MM/dd/yyyy, h:mm:ss a';
export const AGENCIES = ['ALA', 'CAIR', 'CC', 'CET', 'IRC', 'PARS'];

// if they answer YES to these questions then they get a full green flag, basically they have already gone through a vetting process
// to recieve those public benefits so we can trust that process and give them green dot
export const fullWaiverQuestionKeys = ['receive_public_benefits'];

// these are keys of questions that are in the "RED FLAG section" but are not actually red flag questions, i know, its confusing
export const questionKeysThatAreNotRedFlagsButInARedFlagQuestionnaire = [
    'male',
    'still_married_to_that_citizen',
    'receive_public_benefits',
    'live_US_18-26_and_are_26-31',
    'selective_service',
    'green_card_through_marriage',
    'speak_basic_english',
    'follow_up_basic_english_50_years_gc_20_years',
    'follow_up_basic_english_55_years_gc_15_years',
    'taxes_payment_plan',
];
export const actualRedFlagQuestionKeys = [
    'contact_with_police',
    'contact-with-police',
    'habitual_alcoholic_drugs',
    'money_from_illegal_gambling',
    'contact_with_immigration_officer',
    'helped_enter_or_entered_US_illegally',
    'married_to_multiple_people_same_time',
    'failed_support_kids_pay_alimony',
    'asylum_travel_back_home_country',
    'deported_removed_excluded_from_US',
    'lied_to_obtain_immigrant_benefit',
    'lied_to_obtain_welfare_benefit',
    'left_US_>6mo_while_LPR',
    'lived_outsideUS>insideUS_since_LPR',
    'owed_taxes_since_LPR',
    'genocide_torture_killing_hurting',
    'served_military_group_against_govt',
    'court-martialed_disciplinced_in_military',
    'associated_terrorist_orgs_gangs',
    'US_citizen_registered_voted',

];
export type RedFlagKey = ArrayElementOf<typeof actualRedFlagQuestionKeys>;

// need this due to values being set as translated and not the YES or no
export const yesValuesTranslated = ['Yes', 'Sí', 'Có', 'Oo', '是', '是', '是', 'Да', 'አዎ', 'نعم', 'بله', 'हाँ', '예', 'هو', 'ਹਾਂ', 'Sim'];
export type yesValuesTranslatedAsTypedString = 'Yes' | 'Sí' | 'Có' | 'Oo' | '是' | '是' | '是' | 'Да' | 'አዎ' | 'نعم' | 'بله' | 'हाँ' | '예' | 'هو' | 'ਹਾਂ' | 'Sim';

