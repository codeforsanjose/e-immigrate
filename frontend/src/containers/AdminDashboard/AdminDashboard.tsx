import React from 'react';
import { sendRequest } from '../../sendRequest/sendRequest';
import { History } from 'history';
import { format } from 'date-fns';
import { apis } from '../../sendRequest/apis';
import { getAuthToken } from '../../utilities/auth_utils';
import { searchArrayObjects } from '../../utilities/search_array';
import './AdminDashboard.css';
import { languageOptions, workshopTitle } from '../../data/LanguageOptions';
import { Navbar } from '../../compositions/Navbar/Navbar';
import { Button } from '../../components/Button/Button';
import SortArrow from '../../data/images/SortArrow.svg';
import { WithEventTarget } from '../../types/WithEventTarget';
import { defaultCompare } from '../../utilities/defaultCompare';
import { useHistory } from 'react-router-dom';
const {
    getQuestionnaireResponse,
    emailQuestionnaireResponse,
    generateResponsesExcel,
    getQuestions,
    agencyAssignURL,
    assignResponseFlag,
    assignEmail,
    deleteQuestionnaireResponse,
} = apis;
const DESCRIPTIVE_TIMESTAMP = 'MM/dd/yyyy, h:mm:ss a';
const AGENCIES = ['ALA', 'CAIR', 'CC', 'CET', 'IRC', 'PARS'];
const questionKeysThatAreNotRedFlagsButInARedFlagQuestionnaire = [
    'male',
    'still_married_to_that_citizen',
    'receive_public_benefits',
    'live_US_18-26_and_are_26-31',
    'selective_service',
    'green_card_through_marriage',
];

function isAgencyObject(value: unknown): value is { agency: string, } {
    if (value == null) return false;
    else if (typeof value !== 'object') return false;
    if (!('agency' in value)) return false;
    return typeof (value.agency) === 'number';
}
type QuestionnaireResponseElement = {
    _id: string;
    responseDownloadedToExcel: boolean;
    questionnaireResponse: Record<string, string>;
    flagOverride?: boolean;
    flag?: boolean;
    agency: string;
    emailSent?: boolean;
    createdAt: number | Date;
    updatedAt: number | Date;
};
export function AdminDashboard() {
    const history = useHistory();
    const [questionnaireResponses, setQuestionnaireResponses] = React.useState<Array<QuestionnaireResponseElement>>([]);
    const [questions, setQuestions] = React.useState<Array<string>>([]);
    const content = { buttonHome: 'Home' };
    const [language, setLanguage] = React.useState<string>('');
    const [flagOrder, setFlagOrder] = React.useState(false);
    const [emailOrder, setEmailOrder] = React.useState(false);
    const [downloadOrder, setDownloadOrder] = React.useState(false);
    const [loading, setLoading] = React.useState(true);
    const [createdOrder, setCreatedOrder] = React.useState(true);
    const [updatedOrder, setUpdatedOrder] = React.useState(true);
    const [filterBy, setFilterBy] = React.useState('full_name');
    const [searchTerm, setSearchTerm] = React.useState('');
    React.useEffect(() => {
        setLoading(true);
        const jwt = getAuthToken();
        if (jwt === null) {
            history.push('/login');
            return;
        }
        else {
            const requestObj = {
                url: getQuestionnaireResponse,
            };
            const headers = {
                Authorization: `Bearer ${jwt}`,
            };
            sendRequest<{
                responses: Array<QuestionnaireResponseElement>;
            }>(requestObj, headers).then((response) => {
                const { responses = [] } = response;
                const updatedResponses = responses
                    .map((item) => {
                        const { questionnaireResponse = {} } = item;
                        const newFlag = (item.flagOverride ?? false)
                            ? item.flag
                            : Object.entries(questionnaireResponse).reduce(
                                (acc, stuff) => {
                                    const [key, value] = stuff;
                                    return !questionKeysThatAreNotRedFlagsButInARedFlagQuestionnaire.includes(
                                        key,
                                    )
                                        ? (value.length > 0) &&
                                            value.toUpperCase() === 'YES'
                                            ? true
                                            : acc
                                        : acc;
                                },
                                false,
                            );
                        return {
                            ...item,
                            selected: false,
                            flag: newFlag,
                        };
                    })
                    .sort((itemA, itemB) => {
                        if (isAgencyObject(itemA) && isAgencyObject(itemB)) {
                            return itemA.agency > itemB.agency ? 1 : -1;
                        }
                        return -1;
                    });
                setLoading(false);
                setQuestionnaireResponses(updatedResponses);
            });
        }
    }, [history]);

    const toggleFlag = React.useCallback((index: number) => {
        const updatedResponses = questionnaireResponses.map(
            (item, responseIndex) => {
                return {
                    ...item,
                    flag: responseIndex === index ? !(item.flag ?? false) : item.flag,
                    flagOverride: true,
                };
            },
        );
        const responseToUpdate = updatedResponses.reduce(
            (accumulator, item, responseIndex) => {
                return responseIndex === index ? item : accumulator;
            },
            {},
        );
        const requestObj = {
            url: assignResponseFlag,
            method: 'POST',
            body: JSON.stringify({
                responsesToUpdate: [responseToUpdate],
            }),
        };
        const jwt = getAuthToken();
        const headers = {
            Authorization: `Bearer ${jwt}`,
        };
        sendRequest(requestObj, headers).then((response) => {
            setQuestionnaireResponses(updatedResponses);
        });
    }, [questionnaireResponses]);
    const resetEmail = React.useCallback((index: number) => {
        const updatedResponses = questionnaireResponses.map(
            (item, responseIndex) => {
                return {
                    ...item,
                    emailSent: responseIndex === index ? false : item.emailSent,
                };
            },
        );
        const responseToUpdate = updatedResponses.reduce(
            (accumulator, item, responseIndex) => {
                return responseIndex === index ? item : accumulator;
            },
            {},
        );
        const requestObj = {
            url: assignEmail,
            method: 'POST',
            body: JSON.stringify({
                responsesToUpdate: [responseToUpdate],
            }),
        };
        const jwt = getAuthToken();
        const headers = {
            Authorization: `Bearer ${jwt}`,
        };
        sendRequest(requestObj, headers).then((response) => {
            setQuestionnaireResponses(updatedResponses);
        });
    }, [questionnaireResponses]);
    const flagOverviewMarkup = React.useMemo(() => {
        return (
            <div className="flag-dashboard-card">
                <h4>Responses</h4>
                <div>
                    <span>Red:</span>
                    <span className="text-red bold">
                        {questionnaireResponses.filter(
                            (response) => response.flag === true,
                        ).length}
                    </span>
                </div>
                <div>
                    Green:{' '}
                    <span className="text-green bold">
                        {questionnaireResponses.filter(
                            (response) => response.flag === false,
                        ).length}
                    </span>
                </div>
                <div>
                    Total:{' '}
                    <span className="bold">
                        {questionnaireResponses.length}
                    </span>
                </div>
            </div>
        );
    }, [questionnaireResponses]);

    const agencyOverviewMarkup = React.useMemo(() => {
        return (
            <div className="agency-dashboard-card">
                <h4>Assigned to</h4>
                <div className="agency-grid">
                    {AGENCIES.map((agency, idx) => {
                        return (
                            <div key={idx} className="agency-row">
                                <div>{agency}</div>
                                <div className="sum text-red bold">
                                    {questionnaireResponses.filter(
                                        (response) => response.agency === agency &&
                                            response.flag === true,
                                    ).length}
                                </div>
                                <div className="sum text-green bold">
                                    {questionnaireResponses.filter(
                                        (response) => response.agency === agency &&
                                            response.flag === false,
                                    ).length}
                                </div>
                                <div className="sum bold">
                                    {questionnaireResponses.filter(
                                        (response) => response.agency === agency,
                                    ).length}
                                </div>
                            </div>
                        );
                    })}
                </div>
            </div>
        );
    }, [questionnaireResponses]);

    const allOptions = React.useMemo(() => {
        const firstOption = (
            <option key="agency-initial" value="">
                Please select
            </option>
        );
        const agenciesOptions = AGENCIES.map((agency, index) => {
            return (
                <option key={`agency-${index}`} value={agency}>
                    {agency}
                </option>
            );
        });
        return [firstOption, agenciesOptions];
    }, []);

    const assignResponseAgency = React.useCallback((e: WithEventTarget<string>, resIndex: number) => {
        const updatedResponses = questionnaireResponses.map((item, index) => {
            return resIndex === index
                ? {
                    ...item,
                    agency: e.target.value,
                }
                : item;
        });
        const selectedResponseForAgency = {
            ...questionnaireResponses[resIndex],
            agency: e.target.value,
        };
        const requestObj = {
            url: agencyAssignURL,
            method: 'POST',
            body: JSON.stringify({
                responsesToEmail: [selectedResponseForAgency],
            }),
        };
        const jwt = getAuthToken();
        const headers = {
            Authorization: `Bearer ${jwt}`,
        };
        sendRequest(requestObj, headers).then((response) => {
            console.log('wow agency success response', response);
            setQuestionnaireResponses(updatedResponses);
        });
    }, [questionnaireResponses]);

    const sortFlags = () => {
        flagOrder
            ? sortAscending('flag', flagOrder, setFlagOrder)
            : sortDescending('flag', flagOrder, setFlagOrder);
    };
    const sortEmail = () => {
        emailOrder
            ? sortAscendingUndefined('emailSent', emailOrder, setEmailOrder)
            : sortDescendingUndefined('emailSent', emailOrder, setEmailOrder);
    };
    const sortDownload = () => {
        downloadOrder
            ? sortAscendingUndefined(
                'responseDownloadedToExcel',
                downloadOrder,
                setDownloadOrder,
            )
            : sortDescendingUndefined(
                'responseDownloadedToExcel',
                downloadOrder,
                setDownloadOrder,
            );
    };
    const sortCreated = () => {
        createdOrder
            ? sortAscending('createdAt', createdOrder, setCreatedOrder)
            : sortDescending('createdAt', createdOrder, setCreatedOrder);
    };
    const sortUpdated = () => {
        updatedOrder
            ? sortAscending('updatedAt', updatedOrder, setUpdatedOrder)
            : sortDescending('updatedAt', updatedOrder, setUpdatedOrder);
    };
    
    type HeaderState = boolean;
    type SetHeaderState = (value: HeaderState) => void;
    type SortFunction = (property: keyof QuestionnaireResponseElement, headerState: HeaderState, setHeaderState: SetHeaderState) => void;
    const sortAscending: SortFunction = (property, headerState, setHeaderState) => {
        const sortedResponses = questionnaireResponses.sort((a, b) => {
            return defaultCompare(a[property], b[property]);
        });
        setQuestionnaireResponses(sortedResponses);
        setHeaderState(!headerState);
    };
    const sortDescending: SortFunction = (property, headerState, setHeaderState) => {
        const sortedResponses = questionnaireResponses.sort((a, b) => {
            return -defaultCompare(a[property], b[property]);
        });
        setQuestionnaireResponses(sortedResponses);
        setHeaderState(!headerState);
    };
    const sortAscendingUndefined: SortFunction = (property, headerState, setHeaderState) => {
        const sortedResponses = questionnaireResponses.sort((a, b) => {
            if ((Boolean(a[property])) && !(Boolean(b[property]))) return -1;
            if (!(Boolean(a[property])) && (Boolean(b[property]))) return 1;
            return 0;
        });
        setQuestionnaireResponses(sortedResponses);
        setHeaderState(!headerState);
    };
    const sortDescendingUndefined: SortFunction = (property, headerState, setHeaderState) => {
        const sortedResponses = questionnaireResponses.sort((a, b) => {
            if ((Boolean(a[property])) && !(Boolean(b[property]))) return 1;
            if (!(Boolean(a[property])) && (Boolean(b[property]))) return -1;
            return 0;
        });
        setQuestionnaireResponses(sortedResponses);
        setHeaderState(!headerState);
    };
    const softDeleteResponse = React.useCallback((resIndex: number) => {
        const confirmBox = window.confirm(
            'Do you really want to delete this questionnaire response?',
        );
        if (!confirmBox) {
            return;
        }

        const response_id = questionnaireResponses[resIndex]._id;
        const updatedResponses = questionnaireResponses.filter(
            (item, index) => resIndex !== index,
        );

        const requestObj = {
            url: deleteQuestionnaireResponse.replace(':id', response_id),
            method: 'PUT',
        };
        const jwt = getAuthToken();
        const headers = {
            Authorization: `Bearer ${jwt}`,
        };
        sendRequest(requestObj, headers)
            .then((response) => {
                setQuestionnaireResponses(updatedResponses);
            })
            .catch((err) => {
                console.log(
                    `error soft-deleting questionnaire response ${response_id}`,
                    err,
                );
            });
    }, [questionnaireResponses]);

    const responsesMarkup = React.useMemo(() => {
        const filterQuestionnaireResponses = searchArrayObjects(
            questionnaireResponses,
            `questionnaireResponse.${filterBy}`,
            searchTerm,
            3,
        );
        return filterQuestionnaireResponses.map((response, index) => {
            const { questionnaireResponse = {} } = response;
            const fullLangText = languageOptions.find(
                (item) => item.code === questionnaireResponse.languageCode,
            );
            const langDisplay = (fullLangText?.englishName) ?? `Unknown  ${questionnaireResponse.languageCode}`;
            const languageMarkupQuestion = (
                <article key={`td-answer-lang-${index}`} className={`answer `}>
                    <span>
                        Lang:
                        {langDisplay}
                    </span>
                </article>
            );
            const policeMarkupQuestion = (
                <article
                    key={`td-answer-police-${index}`}
                    className={`answer contact-with-police`}
                >
                    <span>
                        contact with police:
                        {questionnaireResponse.contact_with_police}
                    </span>
                </article>
            );
            const policeExplinationMarkupQuestion = (questionnaireResponse.contact_with_police_explanation.length > 0)
                ? (
                    <article
                        key={`td-answer-police-exp-${index}`}
                        className={`answer  contact-with-police-explain`}
                    >
                        <span>
                        police explination:
                            {questionnaireResponse.contact_with_police_explanation}
                        </span>
                    </article>
                )
                : null;
            const alreadyQuestionKeyMarkupedUp = [
                'languageCode',
                'contact_with_police',
                'contact_with_police_explanation',
            ];
            const allAnswers = Object.keys(questionnaireResponse).reduce<Array<JSX.Element>>((accumulator, questionKey, index) => {
                const flagIt = !questionKeysThatAreNotRedFlagsButInARedFlagQuestionnaire.includes(
                    questionKey,
                )
                    ? (questionnaireResponse[questionKey].length > 0) &&
                            questionnaireResponse[questionKey].toUpperCase() ===
                            'YES'
                        ? 'red-outline'
                        : 'green-outline'
                    : 'green-outline';
                const answerMarkup = !alreadyQuestionKeyMarkupedUp.includes(
                    questionKey,
                )
                    ? (
                        <article
                            key={`td-answer-${index}`}
                            className={`answer ${flagIt}`}
                        >
                            <b>{index + 1}.</b>
                            <span>
                                {questionKey}:
                                {questionnaireResponse[questionKey]}
                            </span>
                        </article>
                    )
                    : null;
                return (answerMarkup != null)
                    ? [...accumulator, answerMarkup]
                    : accumulator;
            }, []);
            const allTheAnswers = [
                languageMarkupQuestion,
                policeMarkupQuestion,
                policeExplinationMarkupQuestion,
                ...allAnswers,
            ];
            return (
                <tr key={response._id}>
                    <td>{index + 1}</td>
                    <td>
                        Created:{' '}
                        {format(
                            new Date(response.createdAt),
                            DESCRIPTIVE_TIMESTAMP,
                        )}
                    </td>
                    <td>
                        Updated:{' '}
                        {format(
                            new Date(response.updatedAt),
                            DESCRIPTIVE_TIMESTAMP,
                        )}
                    </td>
                    <td>
                        <div
                            className={`flag ${(response.flag ?? false) ? 'red' : 'green'}`}
                            onClick={(e) => toggleFlag(index)}
                        ></div>
                    </td>
                    <td>
                        <label htmlFor="agency-select">Assigned To:</label>
                        <select
                            id="agency-select"
                            value={response.agency}
                            onChange={(e) => assignResponseAgency(e, index)}
                        >
                            {allOptions}
                        </select>
                    </td>
                    <td>
                        <span>{(response.emailSent ?? false) ? 'Yes' : 'No'}</span>
                        <button onClick={(e) => resetEmail(index)}>
                            RESET
                        </button>
                    </td>
                    <td>
                        <span>
                            {response.responseDownloadedToExcel ? 'Yes' : 'No'}
                        </span>
                    </td>
                    <td>
                        <div
                            title="Delete this response"
                            className="delete"
                            onClick={(e) => softDeleteResponse(index)}
                        ></div>
                    </td>
                    <td>
                        <div className="all-answers">{allTheAnswers}</div>
                    </td>
                </tr>
            );
        });
    }, [questionnaireResponses, filterBy, searchTerm, allOptions, toggleFlag, assignResponseAgency, resetEmail, softDeleteResponse]);

    const responsesTable = (
        <table className="responses">
            <tbody>
                <tr className="header-row">
                    <th>#</th>
                    <th>
                        Created
                        <SortArrow
                            className={createdOrder ? '' : 'up'}
                            onClick={sortCreated} />
                    </th>
                    <th>
                        Updated
                        <SortArrow
                            className={updatedOrder ? '' : 'up'}
                            onClick={sortUpdated} />
                    </th>
                    <th>
                        Flag
                        <SortArrow
                            className={flagOrder ? 'up' : ''}
                            onClick={sortFlags} />
                    </th>
                    <th>Agency</th>
                    <th>
                        Email Sent
                        <SortArrow
                            className={emailOrder ? 'up' : ''}
                            onClick={sortEmail} />
                    </th>
                    <th>
                        Response Downloaded
                        <SortArrow
                            className={downloadOrder ? 'up' : ''}
                            onClick={sortDownload} />
                    </th>
                    <th></th>
                    <th>Responses</th>
                </tr>
                {responsesMarkup}
            </tbody>
        </table>
    );

    const sendEmailsToUsers = () => {
        const requestObj = {
            url: emailQuestionnaireResponse,
            method: 'POST',
            body: JSON.stringify({
                responsesToEmail: [],
            }),
        };
        const jwt = getAuthToken();
        const headers = {
            Authorization: `Bearer ${jwt}`,
        };
        setLoading(true);
        sendRequest(requestObj, headers)
            .then((response) => {
                console.log('emails sent?', response);

                setLoading(false);
                window.location.reload();
            })
            .catch((errors) => {
                setLoading(false);
                console.log(
                    'send emails to users failed errors is',
                    JSON.stringify(errors),
                );
            });
    };

    const getExcelFile = () => {
        setLoading(false);
        window.location.href =
            '/api/generateExcel/getLatest/ResponsesExcel.xlsx';
    };

    const downloadLatestResponsesExcel = () => {
        setLoading(true);
        const includedResponses = questionnaireResponses
            .sort((a, b) => {
                if (a.agency < b.agency) return -1;
                if (a.agency > b.agency) return 1;
                return 0;
            })
            .filter((item) => !item.responseDownloadedToExcel);
        const requestObj = {
            url: generateResponsesExcel,
            method: 'POST',
            body: JSON.stringify({
                questions,
                responses: [],
                downloadAll: false,
            }),
        };
        const jwt = getAuthToken();
        const headers = {
            Authorization: `Bearer ${jwt}`,
        };
        sendRequest(requestObj, headers)
            .then((response) => {
                getExcelFile();
            })
            .catch((errors) => {
                console.log(
                    'Error writing the exel sheet',
                    JSON.stringify(errors),
                );
                alert(
                    'Error writing the exel sheet please contact administrator.',
                );
            });
    };
    const downloadAllResponsesExcel = () => {
        const includedResponses = questionnaireResponses.sort((a, b) => {
            if (a.agency < b.agency) return -1;
            if (a.agency > b.agency) return 1;
            return 0;
        });
        const requestObj = {
            url: generateResponsesExcel,
            method: 'POST',
            body: JSON.stringify({
                questions,
                responses: [],
                downloadAll: true,
            }),
        };
        const jwt = getAuthToken();
        const headers = {
            Authorization: `Bearer ${jwt}`,
        };
        sendRequest(requestObj, headers)
            .then((response) => {
                getExcelFile();
            })
            .catch((errors) => {
                console.log(
                    'Error writing the exel sheet',
                    JSON.stringify(errors),
                );
                alert(
                    'Error writing the exel sheet please contact administrator.',
                );
            });
    };

    React.useEffect(() => {
        const requestObj = {
            url: `${getQuestions}/${workshopTitle}.en`,
        };
        sendRequest<{ questions: Array<string>, }>(requestObj).then((response) => {
            setQuestions(response.questions);
        });
    }, []);

    const loadingMarkup = loading
        ? (
            <div className="loading is-vcentered">Loading...</div>
        )
        : null;

    return (
        <section>
            <Navbar 
                content={content} 
                dashboard={true} 
                language={language}
                setLanguage={setLanguage}
            />
            <section className="AdminDashboard">
                {loadingMarkup}
                <section className="overview-container">
                    <article>
                        <h2 className="dashboard-section-title">Overview</h2>
                        <div className="dashboard-card">
                            {flagOverviewMarkup}
                        </div>
                    </article>
                    <article>
                        <h2 className="dashboard-section-title">By Agency</h2>
                        <div className="dashboard-card">
                            {agencyOverviewMarkup}
                        </div>
                    </article>
                </section>
                <section className="dashboard-buttons-container">
                    <Button label="Send Email" onClick={sendEmailsToUsers} />
                    <Button
                        label="Download Latest Excel"
                        onClick={downloadLatestResponsesExcel} />
                    <Button
                        label="Download All Excel"
                        onClick={downloadAllResponsesExcel} />
                </section>
                <section></section>
                <section className="dashboard-section-title">
                    <input
                        placeholder="Search"
                        className="input-box"
                        onChange={(e) => setSearchTerm(e.target.value)}
                        type="text"
                        name="name" />
                    <select
                        className="dropdown-select"
                        value={filterBy}
                        onChange={(e) => setFilterBy(e.target.value)}
                        name="cars"
                        id="cars"
                    >
                        <option value="full_name">Name</option>
                        <option value="email">Email</option>
                        <option value="mobile_phone">Phone</option>
                        <option value="US_zipcode">Zip</option>
                    </select>
                </section>
                <section>
                    <h2 className="dashboard-section-title">Details</h2>
                    {responsesTable}
                </section>
            </section>
        </section>
    );
}
