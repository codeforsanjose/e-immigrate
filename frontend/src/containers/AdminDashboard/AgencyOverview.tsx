import React from 'react';
import { QuestionnaireResponseElement } from './types';
import { AGENCIES } from './constants';
type AgencyOverviewProps = {
    questionnaireResponses: Array<QuestionnaireResponseElement>;
};
export function AgencyOverview(props: AgencyOverviewProps) {
    const {
        questionnaireResponses,
    } = props;
    return (
        <div className="agency-dashboard-card">
            <h4>Assigned to</h4>
            <div className="agency-grid">
                {AGENCIES.map((agency, idx) => {
                    const responsesForAgency = questionnaireResponses.filter(response => response.agency === agency);
                    return (
                        <div key={idx} className="agency-row">
                            <div>{agency}</div>
                            <div className="sum text-red bold">
                                {responsesForAgency.filter(response => response.flag === true).length}
                            </div>
                            <div className="sum text-green bold">
                                {responsesForAgency.filter(response => response.flag === false).length}
                            </div>
                            <div className="sum bold">
                                {responsesForAgency.length}
                            </div>
                        </div>
                    );
                })}
            </div>
        </div>
    );
}
