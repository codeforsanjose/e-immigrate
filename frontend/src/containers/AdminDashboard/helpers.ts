import { yesValuesTranslated } from "./constants";

type AgencyContainer = {
    agency: string;
};
export function isAgencyObject(value: unknown): value is AgencyContainer {
    if (value == null) return false;
    else if (typeof value !== 'object') return false;
    if (!('agency' in value)) return false;
    return typeof (value.agency) === 'number';
}


export function isValueYes(value: string | null): value is 'Yes' {
    if (value == null) return false;
    return value.length > 0 && value.toUpperCase() === 'YES';
}

export function isValueTranslatedYes(value: string | null) {
    if (value == null) return false;
    return value.length > 0 && yesValuesTranslated.includes(value);
}