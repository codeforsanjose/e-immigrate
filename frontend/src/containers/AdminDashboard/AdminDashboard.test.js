import React from 'react';
import AdminDashboard from './AdminDashboard';

import renderer from 'react-test-renderer';

it('AdminDashboard renders correctly', () => {
    const tree = renderer.create(<AdminDashboard />).toJSON();
    expect(tree).toMatchSnapshot();
});
