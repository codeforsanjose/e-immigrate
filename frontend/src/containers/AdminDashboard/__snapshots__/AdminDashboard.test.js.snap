// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdminDashboard renders correctly 1`] = `
<section
  className="AdminDashboard"
>
  <article
    className="overview-container"
  >
    <h3>
      Overview
    </h3>
    <section
      className="details-container"
    >
      total Responses: 
      0
    </section>
  </article>
  <section>
    <button
      onClick={[Function]}
    >
      Send Email
    </button>
  </section>
  <section>
    <h3>
      Details
    </h3>
    <table
      className="responses"
    >
      <tbody />
    </table>
  </section>
</section>
`;
